src/slurm_plugin/fleet_manager.py

```diff
@@ -0,0 +1,386 @@
+        return False
+
+    def __repr__(self):
+        attrs = ", ".join(["{key}={value}".format(key=key, value=repr(value)) for key, value in self.__dict__.items()])
+        return "{class_name}({attrs})".format(class_name=self.__class__.__name__, attrs=attrs)
+
+    def __str__(self):
+        return f"{self.id}"
+
+    def __hash__(self):
+        return hash(self.id)
+
+    @staticmethod
+    def from_describe_instance_data(instance_info):
+        try:
+            return EC2Instance(
+                instance_info["InstanceId"],
+                instance_info["PrivateIpAddress"],
+                instance_info["PrivateDnsName"].split(".")[0],
+                instance_info["LaunchTime"],
+            )
+        except KeyError as e:
+            logger.error("Unable to retrieve EC2 instance info: %s", e)
+            raise e
+
+
+class FleetManagerException(Exception):
+    """Represent an error during the execution of an action with the FleetManager or FleetManagerFactory."""
+
+    def __init__(self, message: str):
+        super().__init__(message)
+
+
+class FleetManagerFactory:
+    @staticmethod
+    def get_manager(cluster_name, region, boto3_config, cluster_config_file, queue, compute_resource, all_or_nothing):
+        try:
+            cluster_config = FleetManagerFactory._load_cluster_config(cluster_config_file)
+
+            queue_config = next(
+                q_config for q_config in cluster_config["Scheduling"]["SlurmQueues"] if q_config["Name"] == queue
+            )
+            compute_resource_config = next(
+                cr_config for cr_config in queue_config["ComputeResources"] if cr_config["Name"] == compute_resource
+            )
+        except FileNotFoundError as e:
+            message = f"Unable to read {cluster_config_file}, {e}"
+            logger.error(message)
+            raise FleetManagerException(message)
+        except StopIteration:
+            message = (
+                f"Unable to found queue {queue} or compute resource {compute_resource} in the {cluster_config_file}"
+            )
+            logger.error(message)
+            raise FleetManagerException(message)
+
+        if compute_resource_config.get("InstanceTypeList"):
+            return Ec2CreateFleetManager(
+                cluster_name,
+                region,
+                boto3_config,
+                queue_config,
+                compute_resource_config,
+                all_or_nothing,
+            )
+        elif compute_resource_config.get("InstanceType"):
+            return Ec2RunInstancesManager(
+                cluster_name,
+                region,
+                boto3_config,
+                queue_config,
+                compute_resource_config,
+                all_or_nothing,
+            )
+        else:
+            raise Exception(
+                f"InstanceTypes or InstanceType field not found in compute resource config: {compute_resource_config}"
```

Comment:
Do you need to print also queue name to be able to identify the compute resource uniquely?
I may be wrong but you can give two compute resources the same name under different queues

Reply:
Changed. InstanceTypes vs InstanceTypeList too.